source/async_stack.cpp

```diff
@@ -39,10 +42,47 @@ extern "C" {
 inline pthread_key_t folly_async_stack_root_tls_key = 0xFFFF'FFFFu;
 }
 
-#endif  // FOLLY_ASYNC_STACK_ROOT_USE_PTHREAD
+#else
+
+#include <vector>
+
+// TODO: expose array of AsyncStackRootHolder
+
+#endif  // ASYNC_STACK_ROOT_USE_PTHREAD
 
 namespace unifex {
 
+#if ASYNC_STACK_ROOT_USE_PTHREAD == 0
+static std::unique_ptr<std::mutex> asyncStackRootsMutex = std::make_unique<std::mutex>();
+static std::vector<void*> asyncStackRoots;
+
+static void addAsyncStackRootHolder(void* holder) noexcept {
+  std::lock_guard<std::mutex> lock(*asyncStackRootsMutex);
+  asyncStackRoots.push_back(holder);
+}
+
+static void removeAsyncStackRootHolder(void* holder) noexcept {
+  if (asyncStackRootsMutex == nullptr) {
```

Comment:
The main thread that will create the vector and the mutex can end up destroying the vector and mutex, before the thread_local ASR is destroyed. In this case, we won't even have a vector of ASRs to fallback to.

An example of this happening is in https://github.com/facebookexperimental/libunifex/blob/main/examples/synchronous_sleep.cpp